jmbitcoin/jmbitcoin/secp256k1_transaction.py

```diff
@@ -9,825 +9,181 @@
-    txobj["ins"][i]["txinwitness"] = [sig, pub]
-    return serialize(txobj)
 
-def signall(tx, priv):
-    # if priv is a dictionary, assume format is
-    # { 'txinhash:txinidx' : privkey }
-    if isinstance(priv, dict):
-        for e, i in enumerate(deserialize(tx)["ins"]):
-            k = priv["%s:%d" % (i["outpoint"]["hash"], i["outpoint"]["index"])]
-            tx = sign(tx, e, k)
-    else:
-        for i in range(len(deserialize(tx)["ins"])):
-            tx = sign(tx, i, priv)
-    return tx
-
-
-def multisign(tx, i, script, pk, amount=None, hashcode=SIGHASH_ALL):
-    """ Tx is assumed to be serialized. The script passed here is
-    the redeemscript, for example the output of mk_multisig_script.
-    pk is the private key, and must be passed in hex.
-    If amount is not None, the output of p2wsh_multisign is returned.
-    What is returned is a single signature.
-    """
-    if isinstance(tx, str):
-        tx = binascii.unhexlify(tx)
-    if isinstance(script, str):
-        script = binascii.unhexlify(script)
-    if amount:
-        return p2wsh_multisign(tx, i, script, pk, amount, hashcode)
-    modtx = signature_form(tx, i, script, hashcode)
-    return ecdsa_tx_sign(modtx, pk, hashcode)
-
-def p2wsh_multisign(tx, i, script, pk, amount, hashcode=SIGHASH_ALL):
-    """ See note to multisign for the value to pass in as `script`.
-    Tx is assumed to be serialized.
-    """
-    modtx = segwit_signature_form(deserialize(tx), i, script, amount,
-                                  hashcode, decoder_func=lambda x: x)
-    return ecdsa_tx_sign(modtx, pk, hashcode)
-
-def apply_p2wsh_multisignatures(tx, i, script, sigs):
-    """Sigs must be passed in as a list, and must be a
-    complete list for this multisig, in the same order
-    as the list of pubkeys when creating the scriptPubKey.
-    """
-    if isinstance(script, str):
-        script = binascii.unhexlify(script)
-    sigs = [binascii.unhexlify(x) if x[:2] == '30' else x for x in sigs]
-    if isinstance(tx, str):
-        return safe_hexlify(apply_p2wsh_multisignatures(
-            binascii.unhexlify(tx), i, script, sigs))
-    txobj = deserialize(tx)
-    txobj["ins"][i]["script"] = ""
-    txobj["ins"][i]["txinwitness"] = [None] + sigs + [script]
-    return serialize(txobj)
-
-def apply_multisignatures(*args):
-    # tx,i,script,sigs OR tx,i,script,sig1,sig2...,sig[n]
-    tx, i, script = args[0], int(args[1]), args[2]
-    sigs = args[3] if isinstance(args[3], list) else list(args[3:])
+    if not amount:
+        # p2pkh only supported here:
+        input_scriptPubKey = pubkey_to_p2pkh_script(pub)
+        sighash = SignatureHash(input_scriptPubKey, tx, i, hashcode)
+        try:
+            sig = ecdsa_raw_sign(sighash, priv, rawmsg=True) + bytes([hashcode])
+        except Exception as e:
+            return return_err(e)
+        tx.vin[i].scriptSig = CScript([sig, pub])
+        # Verify the signature worked.
+        try:
```

Comment:
Might be a good idea to use `bitcointx.core.bitcoinconsensus.ConsensusVerifyScript` (which uses libbitcoinconsensus). But will require libbitcoinconsensus library to be available (if cannot be found in standard paths, it can be manually loaded with `bitcointx.core.bitcoinconsensus.load_bitcoinconsensus_library()` and then the handle supplied to `ConsensusVerifyScript(..., consensus_library_handle=handle)`

Reply:
Deferring this; in the past we have considered the issue of full consensus validation but because coinjoin is atomic (i.e. one transaction) the risk in broadcasting out-of-consensus transactions is only inconvenience. Would be open to counterarguments.

Reply:
No counterarguments from me -- I just wanted to note that consensus-compatible verification option exists